tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
+            diskPos = lo;
+        }
+
+        // Read until the buffer is full or EOF.
         int cnt = 0;
         int rem = this.buff.length;
         while (rem > 0) {
-            int n = super.read(this.buff, cnt, rem); 
-            if (n < 0) break;
+            int n = super.read(this.buff, cnt, rem);
+            if (n < 0) {
+                break; // EOF
+            }
             cnt += n;
             rem -= n;
         }
-        this.hitEOF = (cnt < this.buff.length);
+
+        // Update "this.diskPos" due to super.read() calls.
         this.diskPos += cnt;
-        return cnt;
+
+        // There is a subtle case to consider as this method exits.  Suppose that:
+        // 1. We didn't fill the buffer because we hit EOF (i.e. length < lo + BuffSz).
+        // 2. The client calls seek() to jump forward, followed by write() to write a byte later in the buffer.
+        //
+        // Then the unset part of buff (represented by "*" below) can actually matter:
+        //
+        //                 length was 4 when fillBuffer() was called
+        //                   v
+        //     buff = [A B C D * * * * ...]
+        //            >>>>>>>>>>>>>> ^
+        //            seek(7)      write(0xFF)
+        //
+        // In that case, the file must be extended to contain the write, and the unset part of the buffer becomes part
+        // of the official file contents.  Fortunately the RandomAccessFile contract gives us an out: when the file is
+        // extended in this way, the unwritten portion can be filled with arbitrary data.  Therefore, we do not need to
+        // do anything to the unset parts of buff.
+        //
+        // It is tempting to offer a slightly stronger contract by zero-filling the unset part of "buff":
+        //
+        //     Arrays.fill(this.buff, cnt, BuffSz, (byte)0);
+        //
+        // But do not be tempted!  That call incurs a performance penalty WITHOUT OFFERING A STRONGER CONTRACT!  There
+        // are still cases where we rely on "super" to extend the file---meaning we inherit the weakest possible
+        // contract from RandomAccessFile.  In particular, this happens when extending the file using setLength() or
+        // when the client makes a giant forward seek followed by a write.
     }
 
-    /* overrides RandomAccessFile.seek(long) */
+    @Override
     public void seek(long pos) throws IOException {
+        this.seeek(pos); // ignore boolean result
+    }
+
+    /**
+     * Extends this.seek(long) by return if a page has been read (used for statistics).
+     *
+     * @return true iff data was read from disk
+     */
+    // NOTE: a call to seeek(this.curr) suffices to restore invariant V2 after this.curr changes.
+    //TODO come up with better name for seeek %)
+    public boolean seeek(long pos) throws IOException {
         // Assert.check(!this.closed);
-        if (pos >= this.hi || pos < this.lo) {
+        boolean didPageRead = false;
```

Comment:
minor, but with the restructure, there's no need for this intermediate variable.  You can set `this.curr` immediately, and then `return true` at the end of the if condition, and `return false` at the end of the method.